File: visitor/oc/bytedesk-oc/ui/controllers/other/BDImageViewController.m
//
//  BDImageViewController.m
//  bytedesk-oc
//
//  Created by 宁金鹏 on 2023/9/13.
//

#import "BDImageViewController.h"
@import Photos;

@interface BDImageViewController ()

@property (nonatomic, strong) UIImageView *imageView;
@property (nonatomic, strong) UIButton *saveButton;

@end

@implementation BDImageViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    NSLog(@"%s", __PRETTY_FUNCTION__);
    self.view.backgroundColor = [UIColor blackColor];
    
    self.imageView = [[UIImageView alloc] initWithFrame:CGRectZero];
    self.imageView.contentMode = UIViewContentModeScaleAspectFit;
    self.imageView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.imageView];
    if (self.image) {
        [self.imageView setImage:self.image];
    }
    
    [NSLayoutConstraint activateConstraints:@[
        [self.imageView.topAnchor constraintEqualToAnchor:self.view.topAnchor],
        [self.imageView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [self.imageView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        [self.imageView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor]
    ]];
    
    self.saveButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.saveButton setTitle:@"保存到相册" forState:UIControlStateNormal];
    self.saveButton.translatesAutoresizingMaskIntoConstraints = NO;
    [self.saveButton addTarget:self action:@selector(saveButtonTapped) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:self.saveButton];
    
    [NSLayoutConstraint activateConstraints:@[
        [self.saveButton.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.saveButton.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor constant:-16]
    ]];
    
    UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(dismissViewController)];
    [self.view addGestureRecognizer:tapGesture];
}

- (void)setImage:(UIImage *)image {
    NSLog(@"%s image:%@", __PRETTY_FUNCTION__, image);
    _image = image;
    [self.imageView setImage:image];
}

- (void)dismissViewController {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)saveButtonTapped {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    if (self.image) {
        [PHPhotoLibrary requestAuthorization:^(PHAuthorizationStatus status) {
            dispatch_async(dispatch_get_main_queue(), ^{
                if (status == PHAuthorizationStatusAuthorized) {
                    [self saveImageToAlbum:self.image];
                } else {
                    [self showAuthorizationAlert];
                }
            });
        }];
    } else {
        NSLog(@"image is nil");
    }
}

- (void)saveImageToAlbum:(UIImage *)image {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^{
        [PHAssetChangeRequest creationRequestForAssetFromImage:image];
    } completionHandler:^(BOOL success, NSError * _Nullable error) {
        dispatch_async(dispatch_get_main_queue(), ^{
            if (success) {
                [self showSaveSuccessAlert];
            } else if (error) {
                NSLog(@"Error saving image to photo album: %@", error.localizedDescription);
                [self showSaveFailureAlert];
            }
        });
    }];
}

- (void)showAuthorizationAlert {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"无保存相册权限" message:@"请授权保存到相册." preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"确定" style:UIAlertActionStyleDefault handler:nil];
    [alert addAction:okAction];
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)showSaveSuccessAlert {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"保存成功" message:@"图片已经成功保存到相册." preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"确定" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self.navigationController dismissViewControllerAnimated:YES completion:nil];
    }];
    [alert addAction:okAction];
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)showSaveFailureAlert {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"保存错误" message:@"保存图片到相册失败，请重试." preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"确定" style:UIAlertActionStyleDefault handler:nil];
    [alert addAction:okAction];
    [self presentViewController:alert animated:YES completion:nil];
}

@end

File: visitor/swift/Sources/bytedesk-swift/ViewController/Other/BDImageViewController.swift
//
//  File.swift
//  
//
//  Created by 宁金鹏 on 2023/9/13.
//

import UIKit
import Photos

// UIScrollViewDelegate // TODO: 增加图片缩放
class BDImageViewController: UIViewController {
    
//    var scrollView: UIScrollView!
//    var imageView: UIImageView!
    
    var image: UIImage? {
        didSet {
            imageView.image = image
            //
//            imageView.sizeToFit()
//            scrollView.contentSize = imageView.bounds.size
//
//            let scrollViewFrame = scrollView.frame
//            let scaleWidth = scrollViewFrame.size.width / scrollView.contentSize.width
//            let scaleHeight = scrollViewFrame.size.height / scrollView.contentSize.height
//            let minScale = min(scaleWidth, scaleHeight)
//
//            scrollView.minimumZoomScale = minScale
//            scrollView.maximumZoomScale = 1.0
//            scrollView.zoomScale = minScale
        }
    }
    
    private let imageView: UIImageView = {
        let imageView = UIImageView()
        imageView.contentMode = .scaleAspectFit
        imageView.translatesAutoresizingMaskIntoConstraints = false
        //
        imageView.backgroundColor = UIColor.red
        return imageView
    }()
    
    private let saveButton: UIButton = {
        let button = UIButton(type: .system)
        button.setTitle("保存到相册", for: .normal)
        button.translatesAutoresizingMaskIntoConstraints = false
        button.addTarget(self, action: #selector(saveButtonTapped), for: .touchUpInside)
        return button
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .black
        
//        scrollView = UIScrollView(frame: view.bounds)
//        scrollView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
//        scrollView.delegate = self
//        scrollView.showsVerticalScrollIndicator = false
//        scrollView.showsHorizontalScrollIndicator = false
//        view.addSubview(scrollView)
        
        view.addSubview(imageView)
//        scrollView.addSubview(imageView)
        NSLayoutConstraint.activate([
            imageView.topAnchor.constraint(equalTo: view.topAnchor),
            imageView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            imageView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            imageView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
        ])
        
        view.addSubview(saveButton)
        NSLayoutConstraint.activate([
            saveButton.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            saveButton.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -16)
        ])
        
//        let doubleTapGesture = UITapGestureRecognizer(target: self, action: #selector(handleDoubleTap(_:)))
//        doubleTapGesture.numberOfTapsRequired = 2
//        scrollView.addGestureRecognizer(doubleTapGesture)
        
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(dismissViewController))
        view.addGestureRecognizer(tapGesture)
        
        self.navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .close, target: self, action: #selector(dismissViewController))
    }
    
    @objc private func dismissViewController() {
//        dismiss(animated: true, completion: nil)
        self.navigationController?.dismiss(animated: true)
    }
    
    @objc private func saveButtonTapped() {
        guard let imageToSave = image else {
            return
        }
        
        PHPhotoLibrary.requestAuthorization { status in
            DispatchQueue.main.async {
                if status == .authorized {
                    self.saveImageToAlbum(image: imageToSave)
                } else {
                    self.showAuthorizationAlert()
                }
            }
        }
    }
    
    private func saveImageToAlbum(image: UIImage) {
        PHPhotoLibrary.shared().performChanges({
            PHAssetChangeRequest.creationRequestForAsset(from: image)
        }, completionHandler: { success, error in
            DispatchQueue.main.async {
                if success {
                    self.showSaveSuccessAlert()
                } else if let error = error {
                    print("Error saving image to photo album: \(error.localizedDescription)")
                    self.showSaveFailureAlert()
                }
            }
        })
    }
    
    private func showAuthorizationAlert() {
        let alert = UIAlertController(title: "无保存相册权限", message: "请授权保存到相册.", preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "确定", style: .default, handler: { _ in
            self.navigationController?.dismiss(animated: true)
        }))
        present(alert, animated: true, completion: nil)
    }
    
    private func showSaveSuccessAlert() {
        let alert = UIAlertController(title: "保存成功", message: "图片已经成功保存到相册.", preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "确定", style: .default, handler: { _ in
            self.navigationController?.dismiss(animated: true)
        }))
        present(alert, animated: true, completion: nil)
    }
    
    private func showSaveFailureAlert() {
        let alert = UIAlertController(title: "保存错误", message: "保存图片到相册失败，请重试", preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "确定", style: .default, handler: { _ in
            self.navigationController?.dismiss(animated: true)
        }))
        present(alert, animated: true, completion: nil)
    }
    
    // MARK: - Gesture Recognizer
        
//    @objc private func handleDoubleTap(_ gestureRecognizer: UITapGestureRecognizer) {
//        if scrollView.zoomScale > scrollView.minimumZoomScale {
//            scrollView.setZoomScale(scrollView.minimumZoomScale, animated: true)
//        } else {
//            let center = gestureRecognizer.location(in: scrollView)
//            let zoomRect = zoomRectForScale(scrollView.maximumZoomScale, center: center)
//            scrollView.zoom(to: zoomRect, animated: true)
//        }
//    }
//
//    private func zoomRectForScale(_ scale: CGFloat, center: CGPoint) -> CGRect {
//        var zoomRect = CGRect.zero
//        zoomRect.size.height = scrollView.frame.size.height / scale
//        zoomRect.size.width = scrollView.frame.size.width / scale
//        let newCenter = imageView.convert(center, from: scrollView)
//        zoomRect.origin.x = newCenter.x - (zoomRect.size.width / 2.0)
//        zoomRect.origin.y = newCenter.y - (zoomRect.size.height / 2.0)
//        return zoomRect
//    }
//
//    // MARK: - UIScrollViewDelegate
//
//    func viewForZooming(in scrollView: UIScrollView) -> UIView? {
//        return imageView
//    }
}
